### PR TITLE
Clean database close attempt

### PR DESCRIPTION
**Pull request by AssemblyJohn**
## Describe your changes

Added a clean database close in the SQL destructor.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [ ] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: lib/ocpp/v201/device_model_storage_sqlite.cpp [hunk 1]
```diff
@@ -20,6 +20,12 @@ DeviceModelStorageSqlite::DeviceModelStorageSqlite(const fs::path& db_path) {
     }
 }
 
+DeviceModelStorageSqlite::~DeviceModelStorageSqlite() {
+    if (sqlite3_close(this->db) != SQLITE_OK) {
```
Comment (marcemmers) [inline lib/ocpp/v201/device_model_storage_sqlite.cpp:24]: I would take a look at how its done in `close_connection_internal` in the `database_connection.cpp` file. You could also finalize all open transactions and use `_v2` close.

Comment (AssemblyJohn) [reply lib/ocpp/v201/device_model_storage_sqlite.cpp:24]: I think the best way forward would be to refactor this a bit and use the database_connection which already takes care of many of the details that you have mentioned.

Comment (marcemmers) [reply lib/ocpp/v201/device_model_storage_sqlite.cpp:24]: Yes I think you are right, just didn't want to blow up the PR like that. In the long run this is better :+1:

**Comment by AssemblyJohn** [comment]
Updated to use the new auto-managed interface.

File: include/ocpp/v201/device_model_storage_sqlite.hpp [hunk 2]
```diff
@@ -26,6 +27,7 @@ class DeviceModelStorageSqlite : public DeviceModelStorage {
     /// \brief Opens SQLite connection at given \p db_path
     /// \param db_path  path to database
     explicit DeviceModelStorageSqlite(const fs::path& db_path);
+    ~DeviceModelStorageSqlite();
```
Comment (marcemmers) [inline include/ocpp/v201/device_model_storage_sqlite.hpp:30]:
```suggestion
    ~DeviceModelStorageSqlite() = default;
```

This should work too right?

**Review by Pietfried — APPROVED** [review]
lgtm

**Comment by christopher-davis-afs** [comment]
The changes here for the `DeviceModelStorageSqlite` class seem to have broken [our tests](https://github.com/EVerest/libocpp/actions/runs/9355354996/job/25750301770?pr=611). It doesn't seem to work with in-memory databases properly?

Are there changes we need to make to [our database setup](https://github.com/EVerest/libocpp/pull/611/files#diff-acd0468b0fc7d05e101c753455aa6e82b0d2efbb9256c1a1fd1891c1e7ca7c40R157-R181)?

**Comment by christopher-davis-afs** [comment]
I found the fix. In `DatabaseConnection::open_connection()` you have the following:

```cpp
    if (this->database_file_path.string().find(":memory:") == std::string::npos and
        !fs::exists(this->database_file_path.parent_path())) {
        fs::create_directories(this->database_file_path.parent_path());
    }
```

[SQLite has multiple ways to initialize in-memory databases](https://www.sqlite.org/inmemorydb.html):

> If two or more distinct but shareable in-memory databases are needed in a single process, then the [mode=memory](https://www.sqlite.org/uri.html#coreqp) query parameter can be used with a [URI filename](https://www.sqlite.org/uri.html) to create a named in-memory database:

The code above only checks for `:memory:` and not `mode=memory`. We need to use `mode=memory` to give the device model database a unique name so we don't mangle the device model db and other dbs. So the code needs to look like this:

```cpp
    if (this->database_file_path.string().find(":memory:") == std::string::npos and
        this->database_file_path.string().find("mode=memory") == std::string::npos and
        !fs::exists(this->database_file_path.parent_path())) {
        fs::create_directories(this->database_file_path.parent_path());
    }
```

**Comment by AssemblyJohn** [comment]
Fix here: https://github.com/EVerest/libocpp/pull/659/files

**Pull request merged**